### ISSUE TITLE
charnames::viacode returning less correct name

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10409]
Migrated from [rt.perl.org#75448](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=75448) (status was 'resolved')

Searchable as RT75448$

**Comment by p5pRT** [comment]
### From @khwilliamson


This is a bug report for perl from khw@&#8203;khw\-desktop\.nonet\,
generated with the help of perlbug 1\.39 running under perl 5\.13\.1\.




---
Unicode occasionally makes mistakes and publishes an incorrect name for 
a character\, such as having a typo\.  That name is grandfathered in as 
legal\, but they later will publish the corrected name which maps to the 
same code point\.

viacode should return the corrected name\.  It doesn't\.


---


---
Flags&#8203;:
&nbsp;     category=library
&nbsp;     severity=medium
&nbsp;     module=charnames


---
Site configuration information for perl 5\.13\.1&#8203;:

Configured by khw at Mon May 24 07&#8203;:06&#8203;:34 MDT 2010\.

Summary of my perl5 \(revision 5 version 13 subversion 1\) configuration&#8203;:
&nbsp;   Commit id&#8203;: b081dd7eaaec2b6ee43335645ab40cff0ca3f91a
&nbsp;   Platform&#8203;:
&nbsp;     osname=linux\, osvers=2\.6\.27\-17\-generic\, archname=i686\-linux
&nbsp;     uname='linux khw\-desktop 2\.6\.27\-17\-generic \#1 smp fri mar 12 
03&#8203;:09&#8203;:00 utc 2010 i686 gnulinux '
&nbsp;     config\_args='\-s \-d \-Dprefix=/home/khw/blead \-Dusedevel 
\-D'optimize=\-g3' \-A'optimize=\-g3' \-A'optimize=\-O0''
&nbsp;     hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;     useithreads=undef\, usemultiplicity=undef
&nbsp;     useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;     use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;     usemymalloc=n\, bincompat5005=undef
&nbsp;   Compiler&#8203;:
&nbsp;     cc='cc'\, ccflags ='\-DDEBUGGING \-fno\-strict\-aliasing \-pipe 
\-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE 
\-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;     optimize='\-O0 \-g3'\,
&nbsp;     cppflags='\-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector 
\-I/usr/local/include'
&nbsp;     ccversion=''\, gccversion='4\.3\.2'\, gccosandvers=''
&nbsp;     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, 
lseeksize=8
&nbsp;     alignbytes=4\, prototype=define
&nbsp;   Linker and Libraries&#8203;:
&nbsp;     ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;     libpth=/usr/local/lib /lib /usr/lib
&nbsp;     libs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;     perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;     libc=/lib/libc\-2\.8\.90\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;     gnulibc\_version='2\.8\.90'
&nbsp;   Dynamic Linking&#8203;:
&nbsp;     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;     cccdlflags='\-fPIC'\, lddlflags='\-shared \-g3 \-g3 \-O0 \-L/usr/local/lib 
\-fstack\-protector'

Locally applied patches&#8203;:




---
@&#8203;INC for perl 5\.13\.1&#8203;:
&nbsp;     lib
&nbsp;     /home/khw/blead/lib/perl5/site\_perl/5\.13\.1/i686\-linux
&nbsp;     /home/khw/blead/lib/perl5/site\_perl/5\.13\.1
&nbsp;     /home/khw/blead/lib/perl5/5\.13\.1/i686\-linux
&nbsp;     /home/khw/blead/lib/perl5/5\.13\.1
&nbsp;     /home/khw/blead/lib/perl5/site\_perl
&nbsp;     \.



---
Environment for perl 5\.13\.1&#8203;:
&nbsp;     HOME=/home/khw
&nbsp;     LANG=en\_US\.UTF\-8
&nbsp;     LANGUAGE \(unset\)
&nbsp;     LD\_LIBRARY\_PATH \(unset\)
&nbsp;     LOGDIR \(unset\)
 
PATH=/home/khw/bin&#8203;:/home/khw/print/bin&#8203;:/bin&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/sbin&#8203;:/usr/games&#8203;:/opt/real/RealPlayer&#8203;:/home/khw/cxoffice/bin
&nbsp;     PERL\_BADLANG \(unset\)
&nbsp;     SHELL=/bin/ksh

**Comment by p5pRT** [comment]
### From @smpeters

On Sun May 30 06&#8203;:43&#8203;:04 2010\, public@&#8203;khwilliamson\.com wrote&#8203;:
> 
> This is a bug report for perl from khw@&#8203;khw\-desktop\.nonet\,
> generated with the help of perlbug 1\.39 running under perl 5\.13\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> Unicode occasionally makes mistakes and publishes an incorrect name
> for
> a character\, such as having a typo\.  That name is grandfathered in as
> legal\, but they later will publish the corrected name which maps to
> the
> same code point\.
> 
> viacode should return the corrected name\.  It doesn't\.
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-

Do you have a example character and name to test with?

Steve

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

Steve Peters via RT wrote&#8203;:
> On Sun May 30 06&#8203;:43&#8203;:04 2010\, public@&#8203;khwilliamson\.com wrote&#8203;:
>> This is a bug report for perl from khw@&#8203;khw\-desktop\.nonet\,
>> generated with the help of perlbug 1\.39 running under perl 5\.13\.1\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> Unicode occasionally makes mistakes and publishes an incorrect name
>> for
>> a character\, such as having a typo\.  That name is grandfathered in as
>> legal\, but they later will publish the corrected name which maps to
>> the
>> same code point\.
>>
>> viacode should return the corrected name\.  It doesn't\.
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \-\-\-
> 
> Do you have a example character and name to test with?
> 
> Steve
> 

Any code point in lib/unicore/NameAliases\.txt will return the less 
correct name\, eg U\+A015\.

I intend to fix this by changing the order of the entries in the table 
generated by mktables\, but haven't gotten around to it\.  I had intended 
to fix it by searching backwards\, but discovered that it was 
significantly slower than a forward search\.  What viacode returns is 
legal in Unicode; it's just less than ideal\, so it is a low priority 
fix\.  We have a lot of behavior that is illegal in Unicode that could be 
fixed first\, and so I'm doing that

I just "took" the ticket\.

And thanks for doing this thankless job of going through bug reports\.

**Comment by p5pRT** [comment]
### From @khwilliamson

See attached\.

**Comment by p5pRT** [comment]
### From @khwilliamson

<details><summary>0001-mktables-Complement-variable-meaning-for-clarity.patch</summary>

```diff
From 28a6cae9263b45608d58739f7c5748bd37424fa2 Mon Sep 17 00:00:00 2001
From: Karl Williamson <public@khwilliamson.com>
Date: Thu, 14 Oct 2010 17:29:12 -0600
Subject: [PATCH] mktables: Complement variable meaning for clarity

When I was designing this code, I struggled for a long time in how to
name the concept the $cdm variable meant.  I was never very happy with
it.  Coming back to it after a year, I realized immediately that the
complement was much easier to understand, so this patch does that and
rewrites the comments to match
---
 lib/unicore/mktables |   67 +++++++++++++++++++++++++------------------------
 1 files changed, 34 insertions(+), 33 deletions(-)

diff --git a/lib/unicore/mktables b/lib/unicore/mktables
index c82d2e4..7431f87 100644
--- a/lib/unicore/mktables
+++ b/lib/unicore/mktables
@@ -3442,14 +3442,14 @@ sub trace { return main::trace(@_); }
         # $i still points to the first potential affected range.  Now find the
         # highest range affected, which will determine the length parameter to
         # splice.  (The input range can span multiple existing ones.)  While
-        # we are looking through the range list, see also if this is an
-        # insertion that will change the values of at least one of the
-        # affected ranges.  We don't need to do this check unless this is an
-        # insertion of non-multiples, and also since this is a boolean, we
-        # don't need to do it if have already determined that it will make a
-        # change; just unconditionally change them.  $cdm is created to be 1
-        # if either of these is true. (The 'c' in the name comes from below)
-        my $cdm = ($operation eq '-' || $replace == $MULTIPLE);
+        # we are looking through the range list, see also if this is a "clean
+        # insertion" that doesn't change the values in any existing range.  It
+        # isn't a clean insertion if it is a deletion, and we know it is a
+        # clean insertion if we are adding duplicates (because by definition
+        # they don't affect existing ranges), so we don't need to do the check
+        # for either of those cases.  But otherwise, start off assuming it is
+        # a clean insertion until proven otherwise.
+        my $clean_insert = ($operation eq '+' && $replace != $MULTIPLE);
         my $j;        # This will point to the highest affected range
 
         # For non-zero types, the standard form is the value itself;
@@ -3462,12 +3462,12 @@ sub trace { return main::trace(@_); }
             # searching
             last if $end < $r->[$j]->start;
 
-            # Here, overlaps the range at $j.  If the value's don't match,
-            # and this is supposedly an insertion, it becomes a change
-            # instead.  This is what the 'c' stands for in $cdm.
-            if (! $cdm) {
+            # Here, overlaps the range at $j.  If the values don't match,
+            # and so far we think this is a clean insertion, it becomes a
+            # non-clean insertion, i.e., a 'change' or 'replace' instead.
+            if ($clean_insert) {
                 if ($r->[$j]->standard_form ne $standard_form) {
-                    $cdm = 1;
+                    $clean_insert = 0;
                 }
                 else {
 
@@ -3481,7 +3481,7 @@ sub trace { return main::trace(@_); }
                         # same, but the non-standardized values aren't.  If
                         # replacing unconditionally, then replace
                         if( $replace == $UNCONDITIONALLY) {
-                            $cdm = 1;
+                            $clean_insert = 0;
                         }
                         else {
 
@@ -3495,13 +3495,13 @@ sub trace { return main::trace(@_); }
                                             && $pre_existing =~ /[a-z]/;
 
                             if ($old_mixed != $new_mixed) {
-                                $cdm = 1 if $new_mixed;
+                                $clean_insert = 0 if $new_mixed;
                                 if (main::DEBUG && $to_trace) {
-                                    if ($cdm) {
-                                        trace "Replacing $pre_existing with $value";
+                                    if ($clean_insert) {
+                                        trace "Retaining $pre_existing over $value";
                                     }
                                     else {
-                                        trace "Retaining $pre_existing over $value";
+                                        trace "Replacing $pre_existing with $value";
                                     }
                                 }
                             }
@@ -3515,13 +3515,13 @@ sub trace { return main::trace(@_); }
                                 my $old_punct = $pre_existing =~ /[-_]/;
 
                                 if ($old_punct != $new_punct) {
-                                    $cdm = 1 if $new_punct;
+                                    $clean_insert = 0 if $new_punct;
                                     if (main::DEBUG && $to_trace) {
-                                        if ($cdm) {
-                                            trace "Replacing $pre_existing with $value";
+                                        if ($clean_insert) {
+                                            trace "Retaining $pre_existing over $value";
                                         }
                                         else {
-                                            trace "Retaining $pre_existing over $value";
+                                            trace "Replacing $pre_existing with $value";
                                         }
                                     }
                                 }   # else existing one is just as "good";
@@ -3607,8 +3607,9 @@ sub trace { return main::trace(@_); }
         }
         else {
 
-            # Here the entire input range is not in the gap before $i.  There
-            # is an affected one, and $j points to the highest such one.
+            # Here part of the input range is not in the gap before $i.  Thus,
+            # there is at least one affected one, and $j points to the highest
+            # such one.
 
             # At this point, here is the situation:
             # This is not an insertion of a multiple, nor of tentative ($NO)
@@ -3624,21 +3625,21 @@ sub trace { return main::trace(@_); }
             #   r[$i-1]->end < $start <= $end <= r[$j]->end
             #
             # Also:
-            #   $cdm is a boolean which is set true if and only if this is a
-            #        change or deletion (multiple was handled above).  In
-            #        other words, it could be renamed to be just $cd.
+            #   $clean_insert is a boolean which is set true if and only if
+            #        this is a "clean insertion", i.e., not a change nor a
+            #        deletion (multiple was handled above).
 
             # We now have enough information to decide if this call is a no-op
-            # or not.  It is a no-op if it is a deletion of a non-existent
-            # range, or an insertion of already existing data.
+            # or not.  It is a no-op if this is an insertion of already
+            # existing data.
 
-            if (main::DEBUG && $to_trace && ! $cdm
+            if (main::DEBUG && $to_trace && $clean_insert
                                          && $i == $j
                                          && $start >= $r->[$i]->start)
             {
                     trace "no-op";
             }
-            return if ! $cdm      # change or delete => not no-op
+            return if $clean_insert
                       && $i == $j # more than one affected range => not no-op
 
                       # Here, r[$i-1]->end < $start <= $end <= r[$i]->end
@@ -3700,7 +3701,7 @@ sub trace { return main::trace(@_); }
                 # Here the new element adds to the one below, but not to the
                 # one above.  If inserting, and only to that one range,  can
                 # just change its ending to include the new one.
-                if ($length == 0 && ! $cdm) {
+                if ($length == 0 && $clean_insert) {
                     $r->[$i-1]->set_end($end);
                     trace "inserted range extends range to below so it is now $r->[$i-1]" if main::DEBUG && $to_trace;
                     return;
@@ -3716,7 +3717,7 @@ sub trace { return main::trace(@_); }
 
                 # Here the new element adds to the one above, but not below.
                 # Mirror the code above
-                if ($length == 0 && ! $cdm) {
+                if ($length == 0 && $clean_insert) {
                     $r->[$j+1]->set_start($start);
                     trace "inserted range extends range to above so it is now $r->[$j+1]" if main::DEBUG && $to_trace;
                     return;
-- 
1.5.6.3


```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

<details><summary>0002--perl-75448-charnames-viacode-returning-less-corr.patch</summary>

```diff
From 892b74ea139f2c905206329b454e851f645d258e Mon Sep 17 00:00:00 2001
From: Karl Williamson <public@khwilliamson.com>
Date: Thu, 14 Oct 2010 19:36:04 -0600
Subject: [PATCH] [perl #75448] charnames::viacode returning less correct name

There are several cases where more than one name is valid for a code
point.  This happens usually when the original name was published with a
typo in it.  It's best for viacode to return the revised name, though
the original remains valid.

The names data is in a table generated by mktables exclusively for
charnames, including vianame (and its kin) and viacode.  The fix is to
mktables to put the more correct name first in the table, so that it is
found first and returned by viacode().

When I originally designed this code, I thought the correct name should
come last in the tables, so someone looping and reading it could just
overwrite the less correct one with the more correct one.

But to save memory  we have the same table shared by viacode and
vianame, and vianame has to recognize both names, so both entries
are needed.  viacode could do an rindex to find the more correct name,
but experiments show that that was twice as slow as going the other
direction.  Therefore, this patch is for speed.

If the tables for vianame and viacode were ever to be split, this patch
could be reverted, if desired, to put things back to the reverse order.
---
 lib/charnames.t      |    8 ++++
 lib/unicore/mktables |  115 +++++++++++++++++++++++++-------------------------
 2 files changed, 66 insertions(+), 57 deletions(-)

diff --git a/lib/charnames.t b/lib/charnames.t
index 4944266..883740e 100644
--- a/lib/charnames.t
+++ b/lib/charnames.t
@@ -960,6 +960,14 @@ is("\N{U+1D0C5}", "\N{BYZANTINE MUSICAL SYMBOL FTHORA SKLIRON CHROMA VASIS}");
         # just read) return the same code point.
         test_vianame($i, $hex, $name);
         test_vianame($i, $hex, $names[$i]);
+
+        # Set up so that a test below of this code point will use the alias
+        # instead of the less-correct original.  We can't test here that
+        # viacode is correct, because the alias file may contain multiple
+        # aliases for the same code point, and viacode should return only the
+        # final one.  So don't do it here; instead rely on the loop below to
+        # pick up the test.
+        $names[$i] = $name;
     }
     close $fh;
 
diff --git a/lib/unicore/mktables b/lib/unicore/mktables
index 7431f87..b7cda64 100644
--- a/lib/unicore/mktables
+++ b/lib/unicore/mktables
@@ -3204,7 +3204,9 @@ sub trace { return main::trace(@_); }
         #                         existing one, but has a different value,
         #                         don't replace the existing one, but insert
         #                         this, one so that the same range can occur
-        #                         multiple times.
+        #                         multiple times.  They are stored LIFO, so
+        #                         that the final one inserted is the first one
+        #                         returned in an ordered search of the table.
         #       => anything else  is the same as => $IF_NOT_EQUIVALENT
         #
         # "same value" means identical for non-type-0 ranges, and it means
@@ -3433,23 +3435,60 @@ sub trace { return main::trace(@_); }
             return;
         }
 
-        # Here, we have taken care of the case where $replace is $NO, which
-        # means that whatever action we now take is done unconditionally.  It
-        # still could be that this call will result in a no-op, if duplicates
-        # aren't allowed, and we are inserting a range that merely duplicates
-        # data already in the range list; or also if deleting a non-existent
-        # range.
-        # $i still points to the first potential affected range.  Now find the
-        # highest range affected, which will determine the length parameter to
-        # splice.  (The input range can span multiple existing ones.)  While
-        # we are looking through the range list, see also if this is a "clean
-        # insertion" that doesn't change the values in any existing range.  It
-        # isn't a clean insertion if it is a deletion, and we know it is a
-        # clean insertion if we are adding duplicates (because by definition
-        # they don't affect existing ranges), so we don't need to do the check
-        # for either of those cases.  But otherwise, start off assuming it is
-        # a clean insertion until proven otherwise.
-        my $clean_insert = ($operation eq '+' && $replace != $MULTIPLE);
+        # Here, we have taken care of the case where $replace is $NO.
+        # Remember that here, r[$i-1]->end < $start <= r[$i]->end
+        # If inserting a multiple record, this is where it goes, before the
+        # first (if any) existing one.  This implies an insertion, and no
+        # change to any existing ranges.  Note that $i can be -1 if this new
+        # range doesn't actually duplicate any existing, and comes at the
+        # beginning of the list.
+        if ($replace == $MULTIPLE) {
+
+            if ($start != $end) {
+                Carp::my_carp_bug("$owner_name_of{$addr}Can't cope with adding a multiple record when the range ($start..$end) contains more than one code point.  No action taken.");
+                return;
+            }
+
+            # Don't add an exact duplicate, as it isn't really a multiple
+            if ($end >= $r->[$i]->start) {
+                if ($r->[$i]->start != $r->[$i]->end) {
+                    Carp::my_carp_bug("$owner_name_of{$addr}Can't cope with adding a multiple record when the other range ($r->[$i]) contains more than one code point.  No action taken.");
+                    return;
+                }
+                return if $value eq $r->[$i]->value && $type eq $r->[$i]->type;
+            }
+
+            trace "Adding multiple record at $i with $start..$end, $value" if main::DEBUG && $to_trace;
+            my @return = splice @$r,
+                                $i,
+                                0,
+                                Range->new($start,
+                                           $end,
+                                           Value => $value,
+                                           Type => $type);
+            if (main::DEBUG && $to_trace) {
+                trace "After splice:";
+                trace 'i-2=[', $i-2, ']', $r->[$i-2] if $i >= 2;
+                trace 'i-1=[', $i-1, ']', $r->[$i-1] if $i >= 1;
+                trace "i  =[", $i, "]", $r->[$i] if $i >= 0;
+                trace 'i+1=[', $i+1, ']', $r->[$i+1] if $i < @$r - 1;
+                trace 'i+2=[', $i+2, ']', $r->[$i+2] if $i < @$r - 2;
+                trace 'i+3=[', $i+3, ']', $r->[$i+3] if $i < @$r - 3;
+            }
+            return @return;
+        }
+
+        # Here, we have taken care of $NO and $MULTIPLE replaces.  This leaves
+        # delete, insert, and replace either unconditionally or if not
+        # equivalent.  $i still points to the first potential affected range.
+        # Now find the highest range affected, which will determine the length
+        # parameter to splice.  (The input range can span multiple existing
+        # ones.)  If this isn't a deletion, while we are looking through the
+        # range list, see also if this is a replacement rather than a clean
+        # insertion; that is if it will change the values of at least one
+        # existing range.  Start off assuming it is an insert, until find it
+        # isn't.
+        my $clean_insert = $operation eq '+';
         my $j;        # This will point to the highest affected range
 
         # For non-zero types, the standard form is the value itself;
@@ -3544,44 +3583,6 @@ sub trace { return main::trace(@_); }
         $j--;        # $j now points to the highest affected range.
         trace "Final affected range is $j: $r->[$j]" if main::DEBUG && $to_trace;
 
-        # If inserting a multiple record, this is where it goes, after all the
-        # existing ones for this range.  This implies an insertion, and no
-        # change to any existing ranges.  Note that $j can be -1 if this new
-        # range doesn't actually duplicate any existing, and comes at the
-        # beginning of the list, in which case we can handle it like any other
-        # insertion, and is easier to do so.
-        if ($replace == $MULTIPLE && $j >= 0) {
-
-            # This restriction could be remedied with a little extra work, but
-            # it won't hopefully ever be necessary
-            if ($r->[$j]->start != $r->[$j]->end) {
-                Carp::my_carp_bug("$owner_name_of{$addr}Can't cope with adding a multiple when the other range ($r->[$j]) contains more than one code point.  No action taken.");
-                return;
-            }
-
-            # Don't add an exact duplicate, as it isn't really a multiple
-            return if $value eq $r->[$j]->value && $type eq $r->[$j]->type;
-
-            trace "Adding multiple record at $j+1 with $start..$end, $value" if main::DEBUG && $to_trace;
-            my @return = splice @$r,
-                                $j+1,
-                                0,
-                                Range->new($start,
-                                           $end,
-                                           Value => $value,
-                                           Type => $type);
-            if (main::DEBUG && $to_trace) {
-                trace "After splice:";
-                trace 'j-2=[', $j-2, ']', $r->[$j-2] if $j >= 2;
-                trace 'j-1=[', $j-1, ']', $r->[$j-1] if $j >= 1;
-                trace "j  =[", $j, "]", $r->[$j] if $j >= 0;
-                trace 'j+1=[', $j+1, ']', $r->[$j+1] if $j < @$r - 1;
-                trace 'j+2=[', $j+2, ']', $r->[$j+2] if $j < @$r - 2;
-                trace 'j+3=[', $j+3, ']', $r->[$j+3] if $j < @$r - 3;
-            }
-            return @return;
-        }
-
         # Here, have taken care of $NO and $MULTIPLE replaces.
         # $j points to the highest affected range.  But it can be < $i or even
         # -1.  These happen only if the insertion is entirely in the gap
@@ -11564,7 +11565,7 @@ sub compile_perl() {
         $alias_sentence = <<END;
 The Name_Alias property adds duplicate code point entries with a corrected
 name.  The original (less correct, but still valid) name will be physically
-first.
+last.
 END
     }
     my $comment;
-- 
1.5.6.3


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Oct 14 19&#8203;:00&#8203;:38 2010\, public@&#8203;khwilliamson\.com wrote&#8203;:
> See attached\.

Thank you\. Applied as 969a34cc81e0088cad802e53b05376b1280dedbb and
53d84487fbdd2060c1a666eacaef6e34ce4a1483\.

Unfortunately\, the bug number was stripped\. I have to fight with git am
to stop it from doing that\. My one\-liner to fix it is flawed&#8203;:

perl \-pe '$\_ = "" if $subject = \(?^Subject&#8203;: \\\[PATCH\[^\]\]\*\\\]\\s\*\(\.\*\)?s\)\[0\];
print "\\n$subject" if $subject and ?^$?'

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Oct 21 06&#8203;:10&#8203;:49 2010\, sprout wrote&#8203;:
> On Thu Oct 14 19&#8203;:00&#8203;:38 2010\, public@&#8203;khwilliamson\.com wrote&#8203;:
> > See attached\.
> 
> Thank you\. Applied as 969a34cc81e0088cad802e53b05376b1280dedbb and
> 53d84487fbdd2060c1a666eacaef6e34ce4a1483\.
> 
> Unfortunately\, the bug number was stripped\. I have to fight with git am
> to stop it from doing that\. My one\-liner to fix it is flawed&#8203;:
> 
> perl \-pe '$\_ = "" if $subject = \(?^Subject&#8203;: \\\[PATCH\[^\]\]\*\\\]\\s\*\(\.\*\)?s\)\[0\];
> print "\\n$subject" if $subject and ?^$?'

Or maybe I forgot to run it through that one\-liner\. :\-\) \(It still has
bugs in it\, though\.\)
